File: java/src/day13/PointOfIncidence.java
package day13;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class PointOfIncidence {
    public static void main(String[] args) {
        try {
            List<String> inputs = getFile("inputs/day13/input.txt");
            System.out.println("Part 1: " + part1(inputs));
            System.out.println("Part 2: " + part2(inputs));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static long part1(List<String> inputs) {
        long result = 0;
        for (String input : inputs) {
            result += 100L * reflectionAcrossHorizontal(input);
            result += reflectionAcrossVertical(input);
        }

        return result;
    }

    private static long part2(List<String> inputs) {
        long result = 0;
        for (String input : inputs) {
            int verLine = 0;
            verLine = 100 * reflectionAcrossHorizontal2(input);
            int horLine = 0;
            horLine = reflectionAcrossVertical2(input);
            result += Math.max(verLine, horLine);
        }
        return result;
    }

    private static int reflectionAcrossVertical(String input) {
        List<String> columns = convertFromRowsToColumns(input);
        return reflectionAcrossLine(columns);
    }

    private static int reflectionAcrossHorizontal(String input) {
        List<String> rows = Arrays.stream(input.trim().split("\n")).toList();
        return reflectionAcrossLine(rows);
    }

    private static int reflectionAcrossVertical2(String input) {
        List<String> columns = convertFromRowsToColumns(input);
        return reflectionAcrossLineWithSmudge(columns);
    }

    private static int reflectionAcrossHorizontal2(String input) {
        List<String> rows = Arrays.stream(input.trim().split("\n")).toList();
        return reflectionAcrossLineWithSmudge(rows);
    }

    private static List<String> convertFromRowsToColumns(String input) {
        List<String> rows = Arrays.stream(input.trim().split("\n")).toList();
        List<List<Character>> columns = new ArrayList<>();
        for (int i = 0; i < rows.get(0).length(); i++) {
            List<Character> column = new ArrayList<>();
            columns.add(column);
        }

        for (int i = 0; i < rows.size(); i++) {
            String row = rows.get(i);
            for (int j = 0; j < row.length(); j++) {
                columns.get(j).add(row.charAt(j));
            }
        }
        List<String> convertToColumns = new ArrayList<>();
        for (List<Character> column : columns) {
            StringBuilder builder = new StringBuilder(column.size());
            for (Character ch : column) {
                builder.append(ch);
            }
            convertToColumns.add(builder.toString());
        }
        return convertToColumns;
    }

    private static int reflectionAcrossLine(List<String> rows) {
        for (int i = 1; i < rows.size(); i++) {
            boolean allEquals = true;
            for (int j = 0; j < i; j++) {
                if ((i + j) >= rows.size()) {
                    break;
                }
                if (!rows.get(i - j - 1).equals(rows.get(i + j))) {
                    allEquals = false;
                    break;
                }
            }
            if (allEquals) {
                return i;
            }
        }
        return 0;
    }

    private static int reflectionAcrossLineWithSmudge(List<String> rows) {
        int line = 0;
        for (int i = 1; i < rows.size(); i++) {
            boolean allEquals = false;
            int mistakesAllowed = 1;
            for (int j = 0; j < i; j++) {
                if ((i + j) >= rows.size()) {
                    break;
                }
                String str1 = rows.get(i - j - 1);
                String str2 = rows.get(i + j);
                if (!str1.equals(str2)) {
                    allEquals = true;
                    for (int z = 0; z < str1.length(); z++) {
                        if (str1.charAt(z) != str2.charAt(z)) {
                            mistakesAllowed--;
                            if (mistakesAllowed < 0) {
                                allEquals = false;
                                break;
                            }
                        }
                    }
                }
            }
            if (allEquals) {
                return i;
            }
        }
        return line;
    }

    private static List<String> getFile(String file) throws IOException {
        String wholeInput = Files.readString(Path.of(file));
        return Arrays.stream(wholeInput.strip().split("\n\n")).toList();
    }
}

File: type-script/day13/point-of-incidence.ts
import * as fs from "fs";

const f: string = fs.readFileSync("../inputs/day13/test_input.txt", { encoding: "utf-8" });
const maps: string[][] = f.split("\n\n").map(x => x.trim().split("\n"));

function part1(maps: string[][]): number {
    let result: number = 0;
    for (let map of maps) {
        result += 100 * reflectAcrossHorizontalLine(map);
        result += reflectAcrossVerticalLine(map);
    }
    return result;
}

function part2(maps: string[][]): number {
    let result: number = 0;
    for (let map of maps) {
        const horLine: number = 100 * reflectAcrossHorizontalLineWithSmudge(map);
        const verLine: number = reflectAcrossVerticalLineWithSmudge(map);
        result += Math.max(horLine, verLine);
    }
    return result;
}

function reflectAcrossHorizontalLine(map: string[]): number {
    for (let i = 1; i < map.length; i++) {
        let allMatch: boolean = true;
        for (let j = 0; j < i; j++) {
            if (i + j >= map.length) {
                break;
            }

            if (map[i + j] !== map[i - j - 1]) {
                allMatch = false;
                break;
            }
        }

        if (allMatch) {
            return i;
        }
    }
    return 0;
}

function reflectAcrossHorizontalLineWithSmudge(map: string[]): number {
    for (let i = 1; i < map.length; i++) {
        let allMatch: boolean = false;
        let mistakes: number = 1;
        for (let j = 0; j < i; j++) {
            if (i + j >= map.length) {
                break;
            }

            const str1: string = map[i + j];
            const str2: string = map[i - j - 1];

            if (str1 !== str2) {
                allMatch = true;
                for (let z = 0; z < str1.length; z++) {
                    if (str1[z] !== str2[z]) {
                        mistakes--;
                    }
                }

                if (mistakes < 0) {
                    allMatch = false;
                    break;
                }
            }
        }

        if (allMatch) {
            return i;
        }
    }
    return 0;
}

function reflectAcrossVerticalLine(map: string[]): number {
    const switchRowsForColumns: string[] = convertRowsToColumns(map);
    return reflectAcrossHorizontalLine(switchRowsForColumns);
}

function reflectAcrossVerticalLineWithSmudge(map: string[]): number {
    const switchRowsForColumns: string[] = convertRowsToColumns(map);
    return reflectAcrossHorizontalLineWithSmudge(switchRowsForColumns);
}

function convertRowsToColumns(map: string[]): string[] {
    const rowsToColumns: string[][] = Array.from({ length: map[0].length }, () => Array(map.length));
    for (let i = 0; i < map.length; i++) {
        for (let j = 0; j < map[0].length; j++) {
            rowsToColumns[j][i] = map[i][j];
        }
    }

    const convert: string[] = rowsToColumns.map(x => x.join(""));
    return convert;
}

console.log("Part 1:", part1(maps));
console.log("Part 2:", part2(maps));
